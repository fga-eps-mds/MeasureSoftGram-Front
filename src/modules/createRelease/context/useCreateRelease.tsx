import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { productQuery } from '@services/product';
import { Changes, Product } from '@customTypes/product';
import { addDays, format } from 'date-fns';
import { mutate } from 'swr';
import { useRouter } from 'next/router';
import { Characteristic, Measure, PreConfigRoot, Subcharacteristic } from '@customTypes/preConfig';
import { CREATE_RELEASE_STEP } from '../consts';

interface CreateReleaseProviderProps {
  children: ReactNode;
  productId: string;
  organizationId: string;
  currentProduct: Product;
}

interface ReleaseInfoForm {
  name: string;
  startDate: string;
  endDate: string;
  characteristics: string[];
  changes: Changes[];
}

interface ConfigPageData {
  characteristicCheckbox: string[];
  setCharacteristicCheckbox: (characteristicCheckbox: string[]) => void;
  subcharacterCheckbox: string[];
  setSubcharacterCheckbox: (subcharacterCheckbox: string[]) => void;
  measureCheckbox: string[];
  setMeasureCheckbox: (measureCheckbox: string[]) => void;
  characteristicData: Characteristic[];
}

interface CreateReleaseContextData {
  releaseInfoForm: ReleaseInfoForm;
  successOnCreation: string;
  preConfigCharacteristics: string[] | undefined;
  productId: string;
  organizationId: string;
  isFirstRelease: boolean;
  configPageData: ConfigPageData;
  // eslint-disable-next-line no-unused-vars
  handleChangeForm: (field: string, value: string | Changes[]) => void;
  // eslint-disable-next-line no-unused-vars
  handleSelectCharacteristics: (characteristic: string) => void;
  createProductReleaseGoal: () => void;
  goToNextStep: () => boolean;
  closeAlert: () => void;
  setCurrentConfig: (data: Characteristic[]) => void;
  changeThreshold: boolean;
  toggleChangeThreshold: () => void;
  getNextStep: (activeStep: number, configPage: number) => number;
  getPreviousStep: (activeStep: number, configPage: number) => number;
  allowChangeConfig: boolean;
  toggleAllowChangeConfig: () => void;
  setUseLastConfig: (value: boolean) => void;
}

const CreateReleaseContext = createContext({} as CreateReleaseContextData);

const defaultEndDate = format(addDays(new Date(), 7), 'yyyy-MM-dd');
const defaulStartDate = format(new Date(), 'yyyy-MM-dd');

export function CreateReleaseProvider({
  children,
  productId,
  organizationId,
  currentProduct
}: CreateReleaseProviderProps) {
  const [preConfigCharacteristics, setPreConfigCharacteristics] = useState<string[]>();
  const [successOnCreation, setSuccessOnCreation] = useState('');
  const [releaseInfoForm, setReleaseInfoForm] = useState<ReleaseInfoForm>({
    endDate: defaultEndDate,
    startDate: defaulStartDate
  } as ReleaseInfoForm);
  const [characteristicData, setCharacteristicData] = useState<Characteristic[]>([]);
  const [characteristicCheckbox, setCharacteristicCheckbox] = useState<string[]>([]);
  const [subcharacterCheckbox, setSubcharacterCheckbox] = useState<string[]>([]);
  const [measureCheckbox, setMeasureCheckbox] = useState<string[]>([]);
  const [changeThreshold, setChangeThreshold] = useState<boolean>(false);
  const [allowChangeConfig, setAllowChangeConfig] = useState<boolean>(false);
  const [useLastConfig, setUseLastConfig] = useState<boolean>(false);
  const [isFirstRelease, setIsFirstRelease] = useState<boolean>(false);

  const router = useRouter();

  async function loadCurrentPreConfig() {
    try {
      const result = await productQuery.getPreConfigEntitiesRelationship(organizationId, productId);
      setPreConfigCharacteristics(result.data.map((item) => item.key));
    } catch (error) {
      console.error(error);
    }
  }

  async function checkForFirstRelease() {
    try {
      const result = await productQuery.getReleaseGoalList(organizationId, productId);
      if (!result.data.length) setIsFirstRelease(true);
    } catch (error) {
      console.error(error);
    }
  }

  async function createProductReleaseGoal() {
    try {
      const data = {
        release_name: releaseInfoForm.name,
        start_at: releaseInfoForm.startDate,
        end_at: releaseInfoForm.endDate,
        changes: releaseInfoForm.changes
      };

      const response = await productQuery.createProductReleaseGoal(organizationId, productId, data);
      await mutate(
        JSON.stringify({ url: `organizations/${organizationId}/products/${productId}/release/`, method: 'get' })
      );
      const releaseId = response?.data?.id;

      if (releaseId) {
        await router.push(
          `/products/${organizationId}-${productId}-${currentProduct.name.toLowerCase()}/releases/${releaseId}`
        );
        setSuccessOnCreation('success');
      }
    } catch (error) {
      setSuccessOnCreation('error');
      console.error(error);
    }
  }

  function handleChangeForm(field: string, value: string | string[] | Changes[]) {
    setReleaseInfoForm((form) => ({
      ...form,
      [field]: value
    }));
  }

  function handleSelectCharacteristics(characteristic: string) {
    const selectedCharacteristics = [...(releaseInfoForm.characteristics || [])];

    if (!selectedCharacteristics.includes(characteristic)) {
      selectedCharacteristics.push(characteristic);
    } else {
      selectedCharacteristics.splice(selectedCharacteristics.indexOf(characteristic), 1);
    }

    handleChangeForm('characteristics', selectedCharacteristics);
  }

  function goToNextStep() {
    const { characteristics, name } = releaseInfoForm;
    return !!characteristics?.length && !!name;
  }

  function closeAlert() {
    setSuccessOnCreation('');
    setReleaseInfoForm({
      endDate: defaultEndDate,
      startDate: defaulStartDate,
      name: '',
      characteristics: [],
      changes: []
    });
  }

  function setCurrentConfig(configData: Characteristic[]) {
    if (configData.length !== characteristicData.length && configData.length) {
      setCharacteristicCheckbox(configData.filter((c => c.weight > 0)).map((c: Characteristic) => c.key))
      const subCharacteristics = configData.flatMap((c: Characteristic) => c.subcharacteristics)
      setSubcharacterCheckbox(subCharacteristics.filter((sc => sc.weight > 0)).map((sc: Subcharacteristic) => sc.key))
      const measures = subCharacteristics.flatMap((sc: Subcharacteristic) => sc.measures)
      setMeasureCheckbox(measures.filter((m => m.weight > 0)).map((m: Measure) => m.key))
    }
    setCharacteristicData(configData);
  }

  async function loadCurrentConfig() {
    try {
      if (currentProduct) {
        const result = (await productQuery.getProductCurrentPreConfig(
          organizationId,
          productId
        )) as unknown as PreConfigRoot;
        setCurrentConfig(result.data.data.characteristics);
      }
    } catch (error) {
      console.error(error);
    }
  }

  function toggleChangeThreshold() {
    setChangeThreshold(!changeThreshold);
  }

  function toggleAllowChangeConfig() {
    setAllowChangeConfig(!allowChangeConfig);
  }

  function getNextStep(activeStep: number, configPage: number) {
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep && configPage === 2) {
      return changeThreshold ? CREATE_RELEASE_STEP.ThresholdStep : CREATE_RELEASE_STEP.ReleaseGoalStep;
    }
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep) {
      return CREATE_RELEASE_STEP.CharacteristicStep;
    }
    return activeStep + 1;
  }

  function getPreviousStep(activeStep: number, configPage: number) {
    if (activeStep === CREATE_RELEASE_STEP.ReleaseGoalStep) {
      if (useLastConfig) {
        setUseLastConfig(false);
        return CREATE_RELEASE_STEP.ReleaseSelectorStep;
      }
      return changeThreshold ? CREATE_RELEASE_STEP.ThresholdStep : CREATE_RELEASE_STEP.CharacteristicStep;
    }
    if (activeStep === CREATE_RELEASE_STEP.CharacteristicStep && configPage !== 0) {
      return CREATE_RELEASE_STEP.CharacteristicStep;
    }
    return activeStep - 1;
  }

  useEffect(() => {
    loadCurrentPreConfig();
    loadCurrentConfig();
    checkForFirstRelease();
  }, [productId]);

  // eslint-disable-next-line react/jsx-no-constructed-context-values
  const value = {
    releaseInfoForm,
    successOnCreation,
    preConfigCharacteristics,
    handleChangeForm,
    handleSelectCharacteristics,
    createProductReleaseGoal,
    goToNextStep,
    closeAlert,
    productId,
    organizationId,
    isFirstRelease,
    setCurrentConfig,
    configPageData: {
      characteristicCheckbox,
      setCharacteristicCheckbox,
      subcharacterCheckbox,
      setSubcharacterCheckbox,
      characteristicData,
      setMeasureCheckbox,
      measureCheckbox
    },
    changeThreshold,
    toggleChangeThreshold,
    getNextStep,
    getPreviousStep,
    allowChangeConfig,
    toggleAllowChangeConfig,
    setUseLastConfig
  };

  return <CreateReleaseContext.Provider value={value}>{children}</CreateReleaseContext.Provider>;
}

export function useCreateReleaseContext(): CreateReleaseContextData {
  return useContext(CreateReleaseContext);
}
