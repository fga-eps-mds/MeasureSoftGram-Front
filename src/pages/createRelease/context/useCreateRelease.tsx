import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { productQuery } from '@services/product';
import { Changes } from '@customTypes/product';
import { addDays, format } from 'date-fns';

interface CreateReleaseProviderProps {
  children: ReactNode;
  productId: string;
  organizationId: string;
}

interface ReleaseInfoForm {
  name: string;
  startDate: string;
  endDate: string;
  characteristics: string[];
  changes: Changes[];
}

interface CreateReleaseContextData {
  releaseInfoForm: ReleaseInfoForm;
  successOnCreation: string;
  preConfigCharacteristics: string[] | undefined;
  // eslint-disable-next-line no-unused-vars
  handleChangeForm: (field: string, value: string | Changes[]) => void;
  // eslint-disable-next-line no-unused-vars
  handleSelectCharacteristics: (characteristic: string) => void;
  createProductReleaseGoal: () => void;
  goToGoalsStep: () => boolean;
  closeAlert: () => void;
}

const CreateReleaseContext = createContext({} as CreateReleaseContextData);

const defaultEndDate = format(addDays(new Date(), 7), 'yyyy-MM-dd');
const defaulStartDate = format(new Date(), 'yyyy-MM-dd');

export function CreateReleaseProvider({ children, productId, organizationId }: CreateReleaseProviderProps) {
  const [preConfigCharacteristics, setPreConfigCharacteristics] = useState<string[]>();
  const [successOnCreation, setSuccessOnCreation] = useState('');
  const [releaseInfoForm, setReleaseInfoForm] = useState<ReleaseInfoForm>({
    endDate: defaultEndDate,
    startDate: defaulStartDate
  } as ReleaseInfoForm);

  async function loadCurrentPreConfig() {
    try {
      const result = await productQuery.getPreConfigEntitiesRelationship(organizationId, productId);
      setPreConfigCharacteristics(result.data.map((item) => item.key));
    } catch (error) {
      console.error(error);
    }
  }

  async function createProductReleaseGoal() {
    try {
      const data = {
        release_name: releaseInfoForm.name,
        start_at: releaseInfoForm.startDate,
        end_at: releaseInfoForm.endDate,
        changes: releaseInfoForm.changes
      };

      await productQuery.createProductReleaseGoal(organizationId, productId, data);

      setSuccessOnCreation('success');
    } catch (error) {
      setSuccessOnCreation('error');
      console.error(error);
    }
  }

  function handleChangeForm(field: string, value: string | string[] | Changes[]) {
    setReleaseInfoForm((form) => ({
      ...form,
      [field]: value
    }));
  }

  function handleSelectCharacteristics(characteristic: string) {
    const selectedCharacteristics = [...(releaseInfoForm.characteristics || [])];

    if (!selectedCharacteristics.includes(characteristic)) {
      selectedCharacteristics.push(characteristic);
    } else {
      selectedCharacteristics.splice(selectedCharacteristics.indexOf(characteristic), 1);
    }

    handleChangeForm('characteristics', selectedCharacteristics);
  }

  function goToGoalsStep() {
    const { characteristics, name } = releaseInfoForm;
    return !!characteristics?.length && !!name;
  }

  function closeAlert() {
    setSuccessOnCreation('');
    setReleaseInfoForm({
      endDate: defaultEndDate,
      startDate: defaulStartDate,
      name: '',
      characteristics: [],
      changes: []
    });
  }

  useEffect(() => {
    loadCurrentPreConfig();
  }, [productId]);

  // eslint-disable-next-line react/jsx-no-constructed-context-values
  const value = {
    releaseInfoForm,
    successOnCreation,
    preConfigCharacteristics,
    handleChangeForm,
    handleSelectCharacteristics,
    createProductReleaseGoal,
    goToGoalsStep,
    closeAlert
  };

  return <CreateReleaseContext.Provider value={value}>{children}</CreateReleaseContext.Provider>;
}

export function useCreateReleaseContext(): CreateReleaseContextData {
  return useContext(CreateReleaseContext);
}
