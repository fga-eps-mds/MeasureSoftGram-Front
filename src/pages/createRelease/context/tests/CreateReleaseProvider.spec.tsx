import '@testing-library/jest-dom';

import React, { useEffect } from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { CreateReleaseProvider, useCreateReleaseContext } from '../useCreateRelease';

jest.mock('@services/product', () => ({
  productQuery: {
    getProductCurrentPreConfig: () => ({data: {"id":25,"name":"3","data":{"characteristics":[{"key":"reliability","weight":50,"subcharacteristics":[{"key":"testing_status","weight":100,"measures":[{"key":"passed_tests","weight":33},{"key":"test_builds","weight":33},{"key":"test_coverage","weight":34}]}]},{"key":"maintainability","weight":50,"subcharacteristics":[{"key":"modifiability","weight":100,"measures":[{"key":"non_complex_file_density","weight":33},{"key":"commented_file_density","weight":33},{"key":"duplication_absense","weight":34}]}]}]},"created_at":"2022-09-15T18:56:57-03:00"}}),
    createProductReleaseGoal: () => ({})
  }
}));

const TestingComponent = () => {
  const {
    handleChangeForm,
    handleSelectCharacteristics,
    closeAlert,
    goToGoalsStep,
    createProductReleaseGoal
  } = useCreateReleaseContext();

  useEffect(() => {
    handleChangeForm('endDate', '30/08/2022')
    handleChangeForm('startDate', '30/08/2022')
    handleChangeForm('name', 'aoba')
    handleSelectCharacteristics('aoba')
  }, [])

  return (
    <>
      <h1>Aoba</h1>
      <button
        type='submit'
        data-testid="close-alert"
        onClick={closeAlert}
      >Fechar alerta</button>
      <button
        type='submit'
        data-testid="goal-step"
        onClick={goToGoalsStep}
      >Denifir pesos</button>
      <button
        type='submit'
        data-testid="create-release-goal"
        onClick={createProductReleaseGoal}
      >Cria metas</button>
    </>
  );
};

describe('<CreateReleaseProvider />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe.skip('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
        >
          <TestingComponent/>
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('Aoba')).toBeInTheDocument()
      })

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Comportamento', () => {
    it('Deve resetar estados', () => {
      const tree = render(
        <CreateReleaseProvider
          organizationId='1'
          productId='3'
        >
          <TestingComponent/>
        </CreateReleaseProvider>
      )

      waitFor(() => {
        expect(tree.getAllByText('Aoba')).toBeInTheDocument()
      })

      fireEvent.click(tree.getByTestId('close-alert'));
      fireEvent.click(tree.getByTestId('goal-step'));
      fireEvent.click(tree.getByTestId('create-release-goal'));
    });
  });
});
