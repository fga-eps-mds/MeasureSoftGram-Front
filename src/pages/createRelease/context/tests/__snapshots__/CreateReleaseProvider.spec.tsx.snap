// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateReleaseProvider /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Aoba
      </h1>
      <button
        data-testid="close-alert"
        type="submit"
      >
        Fechar alerta
      </button>
      <button
        data-testid="goal-step"
        type="submit"
      >
        Denifir pesos
      </button>
      <button
        data-testid="create-release-goal"
        type="submit"
      >
        Cria metas
      </button>
    </div>
  </body>,
  "container": <div>
    <h1>
      Aoba
    </h1>
    <button
      data-testid="close-alert"
      type="submit"
    >
      Fechar alerta
    </button>
    <button
      data-testid="goal-step"
      type="submit"
    >
      Denifir pesos
    </button>
    <button
      data-testid="create-release-goal"
      type="submit"
    >
      Cria metas
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
